<!DOCTYPE html>
<!-- saved from url=(0144)https://socs-stats.ucsc.edu:8787/s/a49a91c4a584b17621169/file_show?path=%2Ftmp%2FRtmpyn99mX%2Fpreview-1bb017bd62e8.dir%2Fdownload-neon-data.html -->
<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">


<meta name="generator" content="pandoc">
<meta http-equiv="X-UA-Compatible" content="IE=EDGE">

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Download NEON Data</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




<script type="text/javascript" src="./download-neon-data_files/MathJax.js"></script></head>

<body>




<h1 class="title toc-ignore">Download NEON Data</h1>



<p>This vignette demonstrates how to use the functions and parameters in this package to download the specific scope NEON soil microbe marker gene sequence data and associated data relevant to your analysis.</p>
<div id="dependencies" class="section level1">
<h1>Dependencies</h1>
<p>To begin, ensure that you have installed all dependencies.</p>
<p>From CRAN:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">install.packages</span>(<span class="st">"dplyr"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">install.packages</span>(<span class="st">"neonUtilities"</span>)</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co">#&gt; Attaching package: 'dplyr'</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">#&gt; The following objects are masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co">#&gt;     filter, lag</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co">#&gt;     intersect, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="kw">library</span>(neonUtilities)</a></code></pre></div>
<p>In addition, source files associated with this package:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">source</span>(<span class="st">"./R/utils.R"</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">source</span>(<span class="st">"./code/params.R"</span>)</a></code></pre></div>
</div>
<div id="set-up-directories" class="section level1">
<h1>Set up directories</h1>
<p>Set up the directory structure associated with the various NEON data products. The structure will depend on what you have written in <code>params.R</code>. For example:</p>
<pre><code>PRESET_OUTDIR = "/data/NEON"
PRESET_OUTDIR_SEQUENCE = "raw_sequence"
PRESET_OUTDIR_SEQMETA = "sequence_metadata"
PRESET_OUTDIR_SOIL = "soil"</code></pre>
<p>This will generate (recursively) the following output directories:</p>
<pre><code>/data/NEON/raw_sequence
/data/NEON/sequence_metadata
/data/NEON/soil</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">makeOutputDirectories</span>()</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co">#&gt; Building output directories from base directory '/data/ZHULAB/NEON_DOB/test_base'.</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#&gt; Using '/data/ZHULAB/NEON_DOB/test_base/Illumina/NEON' as sequence subdirectory.</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">#&gt; Using '/data/ZHULAB/NEON_DOB/test_base/sequence_metadata' as sequence metadata subdirectory.</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="co">#&gt; Using '/data/ZHULAB/NEON_DOB/test_base/soil' as soil data subdirectory.</span></a></code></pre></div>
</div>
<div id="download-data" class="section level1">
<h1>Download data</h1>
<p>All of the downloading functions will refer to the subsetting parameters in <code>params.R</code> by default. These parameters may look like the following:</p>
<pre><code>PRESET_SITES = "all" # Default "all". Can also be character vector of 4-letter NEON site codes e.g. c('ONAQ','RMNP')
PRESET_START_YR_MO = "2016-01"
PRESET_END_YR_MO = format(Sys.Date(), "%Y-%m")
TARGET_GENE = "16S" # Must be "ITS", "16S", or "all"
SEQUENCING_RUNS = c('C25G9', 'B69PP') # Default "all". Can also be character vector of 5-letter NEON sequencing run IDs e.g. c('B69RF','BTP4N')</code></pre>
<p>If time and space are not issues for you, you can download the entire dataset over the course of a few hours. However, these subsetting parameters can be used to specify which parts of the data you are interested in keeping for your analysis.</p>
<p>In addition, the following parameters will be used with the downloading functions by default:</p>
<pre><code>PRESET_CHECK_FILE_SIZE = FALSE
PRESET_RETURN_DATA = TRUE</code></pre>
<p><code>PRESET_CHECK_FILE_SIZE</code> determines whether the script will give you a warning with an estimated file size before downloading a file. <code>PRESET_RETURN_DATA</code> determines whether the downloading functions return data as an R object, or simply downloads data to the specified output directory and returns no value.</p>
<p>You can change any of these parameters on-the-fly by entering different arguments to the downloading functions. For more info, see the help pages for <code>downloadSequenceMetadata()</code>, <code>downloadRawSequenceData()</code>, and <code>downloadRawSoilData()</code>.</p>
<div id="download-metadata" class="section level2">
<h2>Download metadata</h2>
<p>The <code>downloadSequenceMetadataRev()</code> function downloads NEON data product DP1.10108 (“Soil microbe marker gene sequences”) and assembles it using the <code>neonUtilities</code> package. This data product contains a number of tables related to accessing raw sequence data. A joined table containing information about the raw data URLs and the marker gene sequencing parameters is returned by this function. In the following example, we run <code>downloadSequenceMetadataRev()</code> with a number of arguments to narrow the range of data to be downloaded. Unspecified arguments will take on the defaults defined in <code>params.R</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">meta &lt;-<span class="st"> </span><span class="kw">downloadSequenceMetadataRev</span>(<span class="dt">startYrMo =</span> <span class="st">"2018-03"</span>, <span class="dt">endYrMo =</span> <span class="st">"2018-07"</span>, </a>
<a class="sourceLine" id="cb9-2" data-line-number="2">                                    <span class="dt">sites =</span> <span class="kw">c</span>(<span class="st">"OSBS"</span>), <span class="dt">targetGene =</span> <span class="st">"16S"</span>)</a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="co">#&gt; ------------------------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="co">#&gt; You have loaded plyr after dplyr - this is likely to cause problems.</span></a>
<a class="sourceLine" id="cb9-5" data-line-number="5"><span class="co">#&gt; If you need functions from both plyr and dplyr, please load plyr first, then dplyr:</span></a>
<a class="sourceLine" id="cb9-6" data-line-number="6"><span class="co">#&gt; library(plyr); library(dplyr)</span></a>
<a class="sourceLine" id="cb9-7" data-line-number="7"><span class="co">#&gt; ------------------------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-9" data-line-number="9"><span class="co">#&gt; Attaching package: 'plyr'</span></a>
<a class="sourceLine" id="cb9-10" data-line-number="10"><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></a>
<a class="sourceLine" id="cb9-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-12" data-line-number="12"><span class="co">#&gt;     arrange, count, desc, failwith, id, mutate, rename, summarise,</span></a>
<a class="sourceLine" id="cb9-13" data-line-number="13"><span class="co">#&gt;     summarize</span></a>
<a class="sourceLine" id="cb9-14" data-line-number="14"><span class="co">#&gt; loading metadata...</span></a>
<a class="sourceLine" id="cb9-15" data-line-number="15"><span class="co">#&gt; Finding available files</span></a>
<a class="sourceLine" id="cb9-16" data-line-number="16"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-17" data-line-number="17">  <span class="op">|</span><span class="st">                                                                            </span></a>
<a class="sourceLine" id="cb9-18" data-line-number="18"><span class="st">  </span><span class="er">|</span><span class="st">                                                                      </span><span class="er">|</span><span class="st">   </span><span class="dv">0</span>%</a>
<a class="sourceLine" id="cb9-19" data-line-number="19">  <span class="op">|</span><span class="st">                                                                            </span></a>
<a class="sourceLine" id="cb9-20" data-line-number="20"><span class="st">  </span><span class="er">|===================================</span><span class="st">                                   </span><span class="er">|</span><span class="st">  </span><span class="dv">50</span>%</a>
<a class="sourceLine" id="cb9-21" data-line-number="21">  <span class="op">|</span><span class="st">                                                                            </span></a>
<a class="sourceLine" id="cb9-22" data-line-number="22"><span class="st">  </span><span class="er">|======================================================================|</span><span class="st"> </span><span class="dv">100</span>%</a>
<a class="sourceLine" id="cb9-23" data-line-number="23"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-24" data-line-number="24"><span class="co">#&gt; Downloading files totaling approximately 71.3 KiB</span></a>
<a class="sourceLine" id="cb9-25" data-line-number="25"><span class="co">#&gt; Downloading 2 files</span></a>
<a class="sourceLine" id="cb9-26" data-line-number="26"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-27" data-line-number="27">  <span class="op">|</span><span class="st">                                                                            </span></a>
<a class="sourceLine" id="cb9-28" data-line-number="28"><span class="st">  </span><span class="er">|</span><span class="st">                                                                      </span><span class="er">|</span><span class="st">   </span><span class="dv">0</span>%</a>
<a class="sourceLine" id="cb9-29" data-line-number="29">  <span class="op">|</span><span class="st">                                                                            </span></a>
<a class="sourceLine" id="cb9-30" data-line-number="30"><span class="st">  </span><span class="er">|======================================================================|</span><span class="st"> </span><span class="dv">100</span>%</a>
<a class="sourceLine" id="cb9-31" data-line-number="31"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-32" data-line-number="32"><span class="co">#&gt; Unpacking zip files using 1 cores.</span></a>
<a class="sourceLine" id="cb9-33" data-line-number="33"><span class="co">#&gt; Stacking operation across a single core.</span></a>
<a class="sourceLine" id="cb9-34" data-line-number="34"><span class="co">#&gt; Stacking table mmg_soilDnaExtraction</span></a>
<a class="sourceLine" id="cb9-35" data-line-number="35"><span class="co">#&gt; Stacking table mmg_soilMarkerGeneSequencing_16S</span></a>
<a class="sourceLine" id="cb9-36" data-line-number="36"><span class="co">#&gt; Stacking table mmg_soilMarkerGeneSequencing_ITS</span></a>
<a class="sourceLine" id="cb9-37" data-line-number="37"><span class="co">#&gt; Stacking table mmg_soilPcrAmplification_16S</span></a>
<a class="sourceLine" id="cb9-38" data-line-number="38"><span class="co">#&gt; Stacking table mmg_soilPcrAmplification_ITS</span></a>
<a class="sourceLine" id="cb9-39" data-line-number="39"><span class="co">#&gt; Stacking table mmg_soilRawDataFiles</span></a>
<a class="sourceLine" id="cb9-40" data-line-number="40"><span class="co">#&gt; Copied the most recent publication of validation file to /stackedFiles</span></a>
<a class="sourceLine" id="cb9-41" data-line-number="41"><span class="co">#&gt; Copied the most recent publication of variable definition file to /stackedFiles</span></a>
<a class="sourceLine" id="cb9-42" data-line-number="42"><span class="co">#&gt; Finished: Stacked 6 data tables and 2 metadata tables!</span></a>
<a class="sourceLine" id="cb9-43" data-line-number="43"><span class="co">#&gt; Stacking took 0.3567479 secs</span></a>
<a class="sourceLine" id="cb9-44" data-line-number="44"><span class="co">#&gt; All unzipped monthly data folders have been removed.</span></a>
<a class="sourceLine" id="cb9-45" data-line-number="45"><span class="co">#&gt; filtering to 16S data</span></a></code></pre></div>
</div>
<div id="download-raw-sequence-data" class="section level2">
<h2>Download raw sequence data</h2>
<p>Now that we have the metadata table loaded into memory, we retrieve a table of unique raw data files and their sequencing run IDs.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">downloaded &lt;-<span class="st"> </span><span class="kw">downloadRawSequenceData</span>(meta)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="co">#&gt; [1] "There are 92 unique (zipped) raw sequence files to download."</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="co">#&gt; Finished downloading /data/ZHULAB/NEON_DOB/test_base/Illumina/NEON/BMI_Plate82WellA12_16S_C5B2R_R1.fastq.gz</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="co">#&gt; Finished downloading /data/ZHULAB/NEON_DOB/test_base/Illumina/NEON/BMI_Plate82WellA12_16S_C5B2R_R2.fastq.gz</span></a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="co">#&gt; Finished downloading /data/ZHULAB/NEON_DOB/test_base/Illumina/NEON/BMI_Plate82WellA11_16S_C5B2R_R1.fastq.gz</span></a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="co">#&gt; Finished downloading /data/ZHULAB/NEON_DOB/test_base/Illumina/NEON/BMI_Plate82WellA11_16S_C5B2R_R2.fastq.gz</span></a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="co">#&gt; Finished downloading /data/ZHULAB/NEON_DOB/test_base/Illumina/NEON/BMI_Plate82WellA4_16S_C5B2R_R2.fastq.gz</span></a></code></pre></div>
</div>
<div id="reorganize-and-rename-the-downloaded-files" class="section level2">
<h2>Reorganize and rename the downloaded files</h2>
<p>Get the names of all downloaded files.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">outdir_sequence &lt;-<span class="st"> </span><span class="kw">file.path</span>(PRESET_OUTDIR, PRESET_OUTDIR_SEQUENCE)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">fn &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path =</span> outdir_sequence, <span class="dt">pattern =</span> <span class="st">"fastq"</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p>The following function untars each file (if necessary), appends the sequencer run ID to the beginning of each filename, and moves files to subfolders within your sequence directory according to their target gene. The terminal directory name for either target gene will be <code>0_raw</code>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">reorganized &lt;-<span class="st"> </span><span class="kw">organizeRawSequenceData</span>(fn, meta)</a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="co">#&gt; Reorganized files can be found in /data/ZHULAB/NEON_DOB/test_base/Illumina/NEON/ITS and /data/ZHULAB/NEON_DOB/test_base/Illumina/NEON/16S.</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="kw">head</span>(reorganized)</a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="co">#&gt; [1] "BMI_Plate80WellG10_16S_C5B2R_R1.fastq.gz"</span></a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="co">#&gt; [2] "BMI_Plate80WellG10_16S_C5B2R_R2.fastq.gz"</span></a>
<a class="sourceLine" id="cb12-6" data-line-number="6"><span class="co">#&gt; [3] "BMI_Plate80WellG11_16S_C5B2R_R1.fastq.gz"</span></a>
<a class="sourceLine" id="cb12-7" data-line-number="7"><span class="co">#&gt; [4] "BMI_Plate80WellG11_16S_C5B2R_R2.fastq.gz"</span></a>
<a class="sourceLine" id="cb12-8" data-line-number="8"><span class="co">#&gt; [5] "BMI_Plate80WellG12_16S_C5B2R_R1.fastq.gz"</span></a>
<a class="sourceLine" id="cb12-9" data-line-number="9"><span class="co">#&gt; [6] "BMI_Plate80WellG12_16S_C5B2R_R2.fastq.gz"</span></a></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>



</body></html>