---
title: "Download NEON Data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{download-neon-data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
knitr::opts_knit$set(
  root.dir = rprojroot::find_rstudio_root_file()
)
```


This vignette demonstrates how to use the functions and parameters in this package to download the specific scope NEON soil microbe marker gene sequence data and associated data relevant to your analysis. 

# Dependencies

To begin, ensure that you have installed all dependencies.

From CRAN:
```{r, eval=FALSE}
install.packages("dplyr")
install.packages("neonUtilities")
```

```{r}
library(dplyr)
library(neonUtilities)
```

In addition, source files associated with this package:

```{r}
source("./R/utils.R")
source("./code/params.R")
```

# Set up directories

Set up the directory structure associated with the various NEON data products. The structure will depend on what you have written in `params.R`. For example:

```
PRESET_OUTDIR = "/data/NEON"
PRESET_OUTDIR_SEQUENCE = "raw_sequence"
PRESET_OUTDIR_SEQMETA = "sequence_metadata"
PRESET_OUTDIR_SOIL = "soil"
```

This will generate (recursively) the following output directories:

```
/data/NEON/raw_sequence
/data/NEON/sequence_metadata
/data/NEON/soil
```

```{r}
makeOutputDirectories()
```

# Download data

All of the downloading functions will refer to the subsetting parameters in `params.R` by default. These parameters may look like the following:

```
PRESET_SITES = "all" # Default "all". Can also be character vector of 4-letter NEON site codes e.g. c('ONAQ','RMNP')
PRESET_START_YR_MO = "2016-01"
PRESET_END_YR_MO = format(Sys.Date(), "%Y-%m")
TARGET_GENE = "16S" # Must be "ITS", "16S", or "all"
SEQUENCING_RUNS = c('C25G9', 'B69PP') # Default "all". Can also be character vector of 5-letter NEON sequencing run IDs e.g. c('B69RF','BTP4N')
```

If time and space are not issues for you, you can download the entire dataset over the course of a few hours. However, these subsetting parameters can be used to specify which parts of the data you are interested in keeping for your analysis.

In addition, the following parameters will be used with the downloading functions by default:

```
PRESET_CHECK_FILE_SIZE = FALSE
PRESET_RETURN_DATA = TRUE
```

`PRESET_CHECK_FILE_SIZE` determines whether the script will give you a warning with an estimated file size before downloading a file. `PRESET_RETURN_DATA` determines whether the downloading functions return data as an R object, or simply downloads data to the specified output directory and returns no value.

You can change any of these parameters on-the-fly by entering different arguments to the downloading functions. For more info, see the help pages for `downloadSequenceMetadata()`, `downloadRawSequenceData()`, and `downloadRawSoilData()`.

## Download metadata

The `downloadSequenceMetadataRev()` function downloads NEON data product DP1.10108 ("Soil microbe marker gene sequences") and assembles it using the `neonUtilities` package. This data product contains a number of tables related to accessing raw sequence data. A joined table containing information about the raw data URLs and the marker gene sequencing parameters is returned by this function. In the following example, we run `downloadSequenceMetadataRev()` with a number of arguments to narrow the range of data to be downloaded. Unspecified arguments will take on the defaults defined in `params.R`.

```{r}
meta <- downloadSequenceMetadataRev(startYrMo = "2018-03", endYrMo = "2018-07", 
                                    sites = c("OSBS"), targetGene = "16S")
```

## Download raw sequence data

Now that we have the metadata table loaded into memory, we retrieve a table of unique raw data files and their sequencing run IDs. 

```{r, message=1:5}
downloaded <- downloadRawSequenceData(meta)
```

## Reorganize and rename the downloaded files

Get the names of all downloaded files.

```{r}
outdir_sequence <- file.path(PRESET_OUTDIR, PRESET_OUTDIR_SEQUENCE)
fn <- list.files(path = outdir_sequence, pattern = "fastq", full.names = TRUE)
```

The following function untars each file (if necessary), appends the sequencer run ID to the beginning of each filename, and moves files to subfolders within your  sequence directory according to their target gene. The terminal directory name for either target gene will be `0_raw`.

```{r}
reorganized <- organizeRawSequenceData(fn, meta)
head(reorganized)
```

