<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>process-16s-sequences</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">process-16s-sequences</h1>



<div id="dependencies" class="section level1">
<h1>Dependencies</h1>
<p>To begin, ensure that you have installed all dependencies.</p>
<p>From BiocManager:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;ShortRead&quot;</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;Biostrings&quot;</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;dada2&quot;</span>)</a></code></pre></div>
<p>From CRAN:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">install.packages</span>(dplyr)</a></code></pre></div>
<p>Load libraries</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">library</span>(ShortRead)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="co">#&gt; Loading required package: BiocGenerics</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co">#&gt; Loading required package: parallel</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co">#&gt; The following objects are masked from 'package:parallel':</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="co">#&gt; The following objects are masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-13" data-line-number="13"><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></a>
<a class="sourceLine" id="cb3-14" data-line-number="14"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb3-15" data-line-number="15"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-16" data-line-number="16"><span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span></a>
<a class="sourceLine" id="cb3-17" data-line-number="17"><span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span></a>
<a class="sourceLine" id="cb3-18" data-line-number="18"><span class="co">#&gt;     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span></a>
<a class="sourceLine" id="cb3-19" data-line-number="19"><span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span></a>
<a class="sourceLine" id="cb3-20" data-line-number="20"><span class="co">#&gt;     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span></a>
<a class="sourceLine" id="cb3-21" data-line-number="21"><span class="co">#&gt;     union, unique, unsplit, which, which.max, which.min</span></a>
<a class="sourceLine" id="cb3-22" data-line-number="22"><span class="co">#&gt; Loading required package: BiocParallel</span></a>
<a class="sourceLine" id="cb3-23" data-line-number="23"><span class="co">#&gt; Loading required package: Biostrings</span></a>
<a class="sourceLine" id="cb3-24" data-line-number="24"><span class="co">#&gt; Loading required package: S4Vectors</span></a>
<a class="sourceLine" id="cb3-25" data-line-number="25"><span class="co">#&gt; Loading required package: stats4</span></a>
<a class="sourceLine" id="cb3-26" data-line-number="26"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-27" data-line-number="27"><span class="co">#&gt; Attaching package: 'S4Vectors'</span></a>
<a class="sourceLine" id="cb3-28" data-line-number="28"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb3-29" data-line-number="29"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-30" data-line-number="30"><span class="co">#&gt;     expand.grid</span></a>
<a class="sourceLine" id="cb3-31" data-line-number="31"><span class="co">#&gt; Loading required package: IRanges</span></a>
<a class="sourceLine" id="cb3-32" data-line-number="32"><span class="co">#&gt; Loading required package: XVector</span></a>
<a class="sourceLine" id="cb3-33" data-line-number="33"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-34" data-line-number="34"><span class="co">#&gt; Attaching package: 'Biostrings'</span></a>
<a class="sourceLine" id="cb3-35" data-line-number="35"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb3-36" data-line-number="36"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-37" data-line-number="37"><span class="co">#&gt;     strsplit</span></a>
<a class="sourceLine" id="cb3-38" data-line-number="38"><span class="co">#&gt; Loading required package: Rsamtools</span></a>
<a class="sourceLine" id="cb3-39" data-line-number="39"><span class="co">#&gt; Loading required package: GenomeInfoDb</span></a>
<a class="sourceLine" id="cb3-40" data-line-number="40"><span class="co">#&gt; Loading required package: GenomicRanges</span></a>
<a class="sourceLine" id="cb3-41" data-line-number="41"><span class="co">#&gt; Loading required package: GenomicAlignments</span></a>
<a class="sourceLine" id="cb3-42" data-line-number="42"><span class="co">#&gt; Loading required package: SummarizedExperiment</span></a>
<a class="sourceLine" id="cb3-43" data-line-number="43"><span class="co">#&gt; Loading required package: Biobase</span></a>
<a class="sourceLine" id="cb3-44" data-line-number="44"><span class="co">#&gt; Welcome to Bioconductor</span></a>
<a class="sourceLine" id="cb3-45" data-line-number="45"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-46" data-line-number="46"><span class="co">#&gt;     Vignettes contain introductory material; view with</span></a>
<a class="sourceLine" id="cb3-47" data-line-number="47"><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></a>
<a class="sourceLine" id="cb3-48" data-line-number="48"><span class="co">#&gt;     'citation(&quot;Biobase&quot;)', and for packages 'citation(&quot;pkgname&quot;)'.</span></a>
<a class="sourceLine" id="cb3-49" data-line-number="49"><span class="co">#&gt; Loading required package: DelayedArray</span></a>
<a class="sourceLine" id="cb3-50" data-line-number="50"><span class="co">#&gt; Loading required package: matrixStats</span></a>
<a class="sourceLine" id="cb3-51" data-line-number="51"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-52" data-line-number="52"><span class="co">#&gt; Attaching package: 'matrixStats'</span></a>
<a class="sourceLine" id="cb3-53" data-line-number="53"><span class="co">#&gt; The following objects are masked from 'package:Biobase':</span></a>
<a class="sourceLine" id="cb3-54" data-line-number="54"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-55" data-line-number="55"><span class="co">#&gt;     anyMissing, rowMedians</span></a>
<a class="sourceLine" id="cb3-56" data-line-number="56"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-57" data-line-number="57"><span class="co">#&gt; Attaching package: 'DelayedArray'</span></a>
<a class="sourceLine" id="cb3-58" data-line-number="58"><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></a>
<a class="sourceLine" id="cb3-59" data-line-number="59"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-60" data-line-number="60"><span class="co">#&gt;     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span></a>
<a class="sourceLine" id="cb3-61" data-line-number="61"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb3-62" data-line-number="62"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-63" data-line-number="63"><span class="co">#&gt;     aperm, apply, rowsum</span></a>
<a class="sourceLine" id="cb3-64" data-line-number="64"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb3-65" data-line-number="65"><span class="kw">library</span>(dada2)</a>
<a class="sourceLine" id="cb3-66" data-line-number="66"><span class="co">#&gt; Loading required package: Rcpp</span></a>
<a class="sourceLine" id="cb3-67" data-line-number="67"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb3-68" data-line-number="68"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-69" data-line-number="69"><span class="co">#&gt; Attaching package: 'dplyr'</span></a>
<a class="sourceLine" id="cb3-70" data-line-number="70"><span class="co">#&gt; The following object is masked from 'package:ShortRead':</span></a>
<a class="sourceLine" id="cb3-71" data-line-number="71"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-72" data-line-number="72"><span class="co">#&gt;     id</span></a>
<a class="sourceLine" id="cb3-73" data-line-number="73"><span class="co">#&gt; The following objects are masked from 'package:GenomicAlignments':</span></a>
<a class="sourceLine" id="cb3-74" data-line-number="74"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-75" data-line-number="75"><span class="co">#&gt;     first, last</span></a>
<a class="sourceLine" id="cb3-76" data-line-number="76"><span class="co">#&gt; The following object is masked from 'package:matrixStats':</span></a>
<a class="sourceLine" id="cb3-77" data-line-number="77"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-78" data-line-number="78"><span class="co">#&gt;     count</span></a>
<a class="sourceLine" id="cb3-79" data-line-number="79"><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></a>
<a class="sourceLine" id="cb3-80" data-line-number="80"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-81" data-line-number="81"><span class="co">#&gt;     combine</span></a>
<a class="sourceLine" id="cb3-82" data-line-number="82"><span class="co">#&gt; The following objects are masked from 'package:GenomicRanges':</span></a>
<a class="sourceLine" id="cb3-83" data-line-number="83"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-84" data-line-number="84"><span class="co">#&gt;     intersect, setdiff, union</span></a>
<a class="sourceLine" id="cb3-85" data-line-number="85"><span class="co">#&gt; The following object is masked from 'package:GenomeInfoDb':</span></a>
<a class="sourceLine" id="cb3-86" data-line-number="86"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-87" data-line-number="87"><span class="co">#&gt;     intersect</span></a>
<a class="sourceLine" id="cb3-88" data-line-number="88"><span class="co">#&gt; The following objects are masked from 'package:Biostrings':</span></a>
<a class="sourceLine" id="cb3-89" data-line-number="89"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-90" data-line-number="90"><span class="co">#&gt;     collapse, intersect, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb3-91" data-line-number="91"><span class="co">#&gt; The following object is masked from 'package:XVector':</span></a>
<a class="sourceLine" id="cb3-92" data-line-number="92"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-93" data-line-number="93"><span class="co">#&gt;     slice</span></a>
<a class="sourceLine" id="cb3-94" data-line-number="94"><span class="co">#&gt; The following objects are masked from 'package:IRanges':</span></a>
<a class="sourceLine" id="cb3-95" data-line-number="95"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-96" data-line-number="96"><span class="co">#&gt;     collapse, desc, intersect, setdiff, slice, union</span></a>
<a class="sourceLine" id="cb3-97" data-line-number="97"><span class="co">#&gt; The following objects are masked from 'package:S4Vectors':</span></a>
<a class="sourceLine" id="cb3-98" data-line-number="98"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-99" data-line-number="99"><span class="co">#&gt;     first, intersect, rename, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb3-100" data-line-number="100"><span class="co">#&gt; The following objects are masked from 'package:BiocGenerics':</span></a>
<a class="sourceLine" id="cb3-101" data-line-number="101"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-102" data-line-number="102"><span class="co">#&gt;     combine, intersect, setdiff, union</span></a>
<a class="sourceLine" id="cb3-103" data-line-number="103"><span class="co">#&gt; The following objects are masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb3-104" data-line-number="104"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-105" data-line-number="105"><span class="co">#&gt;     filter, lag</span></a>
<a class="sourceLine" id="cb3-106" data-line-number="106"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb3-107" data-line-number="107"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-108" data-line-number="108"><span class="co">#&gt;     intersect, setdiff, setequal, union</span></a></code></pre></div>
<p>In addition, source files associated with this package:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw">source</span>(<span class="st">&quot;./R/utils.R&quot;</span>)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="kw">source</span>(<span class="st">&quot;./code/params.R&quot;</span>)</a></code></pre></div>
</div>
<div id="setup-variables" class="section level1">
<h1>Setup variables</h1>
<p>Get filepath names:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="cf">if</span>(<span class="kw">is.null</span>(PRESET_OUTDIR_SEQUENCE) <span class="op">|</span><span class="st"> </span>PRESET_OUTDIR_SEQUENCE <span class="op">==</span><span class="st"> &quot;&quot;</span>) {</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">  PATH_16S &lt;-<span class="st"> </span><span class="kw">file.path</span>(PRESET_OUTDIR, <span class="st">&quot;raw_sequence&quot;</span>, <span class="st">&quot;16S&quot;</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">} <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">  PATH_16S &lt;-<span class="st"> </span><span class="kw">file.path</span>(PRESET_OUTDIR, PRESET_OUTDIR_SEQUENCE, <span class="st">&quot;16S&quot;</span>)</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">}</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">PATH_RAW &lt;-<span class="st"> </span><span class="kw">file.path</span>(PATH_16S, <span class="st">&quot;0_raw&quot;</span>)</a>
<a class="sourceLine" id="cb5-7" data-line-number="7">PATH_TRIMMED &lt;-<span class="st"> </span><span class="kw">file.path</span>(PATH_16S, <span class="st">&quot;1_trimmed&quot;</span>)</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">PATH_FILTERED &lt;-<span class="st"> </span><span class="kw">file.path</span>(PATH_16S, <span class="st">&quot;2_filtered&quot;</span>)</a>
<a class="sourceLine" id="cb5-9" data-line-number="9">PATH_SEQTABS &lt;-<span class="st"> </span><span class="kw">file.path</span>(PATH_16S, <span class="st">&quot;3_seqtabs&quot;</span>)</a>
<a class="sourceLine" id="cb5-10" data-line-number="10">PATH_TRACK &lt;-<span class="st"> </span><span class="kw">file.path</span>(PATH_16S, <span class="st">&quot;track_reads&quot;</span>)</a></code></pre></div>
<p>Get all run IDs so you can group by them:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">unique_runs &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">unlist</span>(</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">  <span class="kw">regmatches</span>(<span class="kw">list.files</span>(PATH_RAW), <span class="kw">gregexpr</span>(<span class="st">&quot;^run[A-Za-z0-9]*&quot;</span>, <span class="kw">list.files</span>(PATH_RAW)))</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">))</a></code></pre></div>
<p>If SMALL_SUBSET == TRUE, run only the first runID. This can be useful for debugging.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="cf">if</span> (SMALL_SUBSET) {</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">  loop_length &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3">} <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">  loop_length &lt;-<span class="st"> </span><span class="kw">length</span>(unique_runs)</a>
<a class="sourceLine" id="cb7-5" data-line-number="5">}</a></code></pre></div>
</div>
<div id="process-reads" class="section level1">
<h1>Process reads</h1>
<p>We use DADA2 to process reads one sequencing run at a time.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">all.seqtabs &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">t1 &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()</a>
<a class="sourceLine" id="cb8-3" data-line-number="3">ti &lt;-<span class="st"> </span><span class="kw">c</span>()</a>
<a class="sourceLine" id="cb8-4" data-line-number="4">first &lt;-<span class="st"> </span><span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>loop_length) {</a>
<a class="sourceLine" id="cb8-6" data-line-number="6">  runID &lt;-<span class="st"> </span>unique_runs[i]</a>
<a class="sourceLine" id="cb8-7" data-line-number="7">  <span class="kw">message</span>(<span class="kw">paste0</span>(<span class="st">&quot;Began processing &quot;</span>, runID, <span class="st">&quot; at &quot;</span>, <span class="kw">Sys.time</span>()))</a>
<a class="sourceLine" id="cb8-8" data-line-number="8"></a>
<a class="sourceLine" id="cb8-9" data-line-number="9">  <span class="co"># Forward and reverse fastq filenames have format: SAMPLENAME_R1.fastq and SAMPLENAME_R2.fastq</span></a>
<a class="sourceLine" id="cb8-10" data-line-number="10">  fnFs &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">list.files</span>(PATH_RAW, <span class="dt">pattern=</span><span class="kw">paste0</span>(runID, <span class="st">&quot;.*_R1.fastq&quot;</span>), <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb8-11" data-line-number="11">  fnRs &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">list.files</span>(PATH_RAW, <span class="dt">pattern=</span><span class="kw">paste0</span>(runID, <span class="st">&quot;.*_R2.fastq&quot;</span>), <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb8-12" data-line-number="12"></a>
<a class="sourceLine" id="cb8-13" data-line-number="13">  <span class="co"># If SMALL_SUBSET == TRUE,</span></a>
<a class="sourceLine" id="cb8-14" data-line-number="14">  <span class="co"># keep only the first two forward-reverse pairs of sequence files</span></a>
<a class="sourceLine" id="cb8-15" data-line-number="15">  <span class="cf">if</span>(SMALL_SUBSET){</a>
<a class="sourceLine" id="cb8-16" data-line-number="16">    <span class="cf">if</span>(<span class="kw">length</span>(fnFs <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)) fnFs &lt;-<span class="st"> </span>fnFs[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</a>
<a class="sourceLine" id="cb8-17" data-line-number="17">    <span class="cf">if</span>(<span class="kw">length</span>(fnRs <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)) fnRs &lt;-<span class="st"> </span>fnRs[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</a>
<a class="sourceLine" id="cb8-18" data-line-number="18">  }</a>
<a class="sourceLine" id="cb8-19" data-line-number="19"></a>
<a class="sourceLine" id="cb8-20" data-line-number="20">  <span class="co"># Remove any files that only have forward or reverse reads</span></a>
<a class="sourceLine" id="cb8-21" data-line-number="21">  matched_fn &lt;-<span class="st"> </span><span class="kw">remove_unmatched_files</span>(fnFs, fnRs)</a>
<a class="sourceLine" id="cb8-22" data-line-number="22">  fnFs &lt;-<span class="st"> </span>matched_fn[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb8-23" data-line-number="23">  fnRs &lt;-<span class="st"> </span>matched_fn[[<span class="dv">2</span>]]</a>
<a class="sourceLine" id="cb8-24" data-line-number="24">  </a>
<a class="sourceLine" id="cb8-25" data-line-number="25">  fn_base &lt;-<span class="st"> </span><span class="kw">basename</span>(<span class="kw">c</span>(fnFs, fnRs))</a>
<a class="sourceLine" id="cb8-26" data-line-number="26"></a>
<a class="sourceLine" id="cb8-27" data-line-number="27">  <span class="co"># Trim reads based on the primer lengths supplied in params.r</span></a>
<a class="sourceLine" id="cb8-28" data-line-number="28">  trim_trackReads &lt;-<span class="st"> </span><span class="kw">trimPrimers16S</span>(fn_base, PATH_RAW, PATH_TRIMMED, <span class="st">&quot;CCTACGGGNBGCASCAG&quot;</span>, <span class="st">&quot;GACTACNVGGGTATCTAATCC&quot;</span>)</a>
<a class="sourceLine" id="cb8-29" data-line-number="29"></a>
<a class="sourceLine" id="cb8-30" data-line-number="30">  <span class="co"># Filter reads based on the settings in params.r</span></a>
<a class="sourceLine" id="cb8-31" data-line-number="31">  filter_trackReads &lt;-<span class="st"> </span><span class="kw">qualityFilter16S</span>(fn_base, PATH_TRIMMED, PATH_FILTERED, MULTITHREAD, MAX_EE_FWD, MAX_EE_REV, <span class="kw">c</span>(<span class="dv">265</span>, <span class="dv">210</span>))</a>
<a class="sourceLine" id="cb8-32" data-line-number="32"></a>
<a class="sourceLine" id="cb8-33" data-line-number="33">  <span class="co"># Now create sequence table for run</span></a>
<a class="sourceLine" id="cb8-34" data-line-number="34">  seqtab.list &lt;-<span class="st"> </span><span class="kw">runDada16S</span>(fn_base, PATH_FILTERED, MULTITHREAD, VERBOSE)</a>
<a class="sourceLine" id="cb8-35" data-line-number="35"></a>
<a class="sourceLine" id="cb8-36" data-line-number="36">  <span class="co"># Create output tracking file</span></a>
<a class="sourceLine" id="cb8-37" data-line-number="37">  track &lt;-<span class="st"> </span><span class="kw">cbind.data.frame</span>(trim_trackReads,</a>
<a class="sourceLine" id="cb8-38" data-line-number="38">                            <span class="dt">filtered =</span> filter_trackReads[,<span class="dv">2</span>],</a>
<a class="sourceLine" id="cb8-39" data-line-number="39">                            seqtab.list<span class="op">$</span>track)</a>
<a class="sourceLine" id="cb8-40" data-line-number="40"></a>
<a class="sourceLine" id="cb8-41" data-line-number="41">  <span class="co"># Append sequence table to output list</span></a>
<a class="sourceLine" id="cb8-42" data-line-number="42">  all.seqtabs[[runID]] &lt;-<span class="st"> </span>seqtab.list<span class="op">$</span>seqtab.nochim</a>
<a class="sourceLine" id="cb8-43" data-line-number="43"></a>
<a class="sourceLine" id="cb8-44" data-line-number="44">  <span class="co"># Save tracking table (which tracks no. of reads remaining at each stage) and sequence table</span></a>
<a class="sourceLine" id="cb8-45" data-line-number="45">  <span class="cf">if</span>(SMALL_SUBSET) {</a>
<a class="sourceLine" id="cb8-46" data-line-number="46">    <span class="kw">write.csv</span>(track, <span class="kw">file.path</span>(PATH_TRACK, <span class="kw">paste0</span>(<span class="st">&quot;track_reads_&quot;</span>,runID,<span class="st">&quot;_SMALLSUBSET.csv&quot;</span>)))</a>
<a class="sourceLine" id="cb8-47" data-line-number="47">    <span class="kw">saveRDS</span>(seqtab.list<span class="op">$</span>seqtab.nochim, <span class="kw">file.path</span>(PATH_SEQTABS, <span class="kw">paste0</span>(<span class="st">&quot;NEON_16S_seqtab_nochim_&quot;</span>, runID, <span class="st">&quot;_SMALLSUBSET.rds&quot;</span>)))</a>
<a class="sourceLine" id="cb8-48" data-line-number="48">  } <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb8-49" data-line-number="49">    <span class="kw">write.csv</span>(track, <span class="kw">file.path</span>(PATH_TRACK, <span class="kw">paste0</span>(<span class="st">&quot;track_reads_&quot;</span>,runID,<span class="st">&quot;.csv&quot;</span>)))</a>
<a class="sourceLine" id="cb8-50" data-line-number="50">    <span class="kw">saveRDS</span>(seqtab.list<span class="op">$</span>seqtab.nochim, <span class="kw">file.path</span>(PATH_SEQTABS, <span class="kw">paste0</span>(<span class="st">&quot;NEON_16S_seqtab_nochim_&quot;</span>, runID, <span class="st">&quot;.rds&quot;</span>)))</a>
<a class="sourceLine" id="cb8-51" data-line-number="51">  }</a>
<a class="sourceLine" id="cb8-52" data-line-number="52">  <span class="kw">message</span>(<span class="kw">paste0</span>(<span class="st">&quot;Finished processing reads in &quot;</span>, runID, <span class="st">&quot; at &quot;</span>, <span class="kw">Sys.time</span>()))</a>
<a class="sourceLine" id="cb8-53" data-line-number="53">  <span class="kw">message</span>(<span class="kw">paste0</span>(<span class="st">&quot;Sequencing run-specific sequence tables can be found in &quot;</span>, PATH_SEQTABS))</a>
<a class="sourceLine" id="cb8-54" data-line-number="54">}</a>
<a class="sourceLine" id="cb8-55" data-line-number="55"><span class="co">#&gt; Began processing runC5B2R at 2020-09-09 18:49:31</span></a>
<a class="sourceLine" id="cb8-56" data-line-number="56"><span class="co">#&gt; 18851570 total bases in 71138 reads from 2 samples will be used for learning the error rates.</span></a>
<a class="sourceLine" id="cb8-57" data-line-number="57"><span class="co">#&gt; 14938980 total bases in 71138 reads from 2 samples will be used for learning the error rates.</span></a>
<a class="sourceLine" id="cb8-58" data-line-number="58"><span class="co">#&gt; Processing: runC5B2R_BMI_Plate77WellA10_16S_C5B2R </span></a>
<a class="sourceLine" id="cb8-59" data-line-number="59"><span class="co">#&gt; Sample 1 - 35659 reads in 26214 unique sequences.</span></a>
<a class="sourceLine" id="cb8-60" data-line-number="60"><span class="co">#&gt; Sample 1 - 35659 reads in 25269 unique sequences.</span></a>
<a class="sourceLine" id="cb8-61" data-line-number="61"><span class="co">#&gt; Duplicate sequences in merged output.</span></a>
<a class="sourceLine" id="cb8-62" data-line-number="62"><span class="co">#&gt; Exact sequence variants inferred for sample: runC5B2R_BMI_Plate77WellA10_16S_C5B2R . </span></a>
<a class="sourceLine" id="cb8-63" data-line-number="63"><span class="co">#&gt; Processing: runC5B2R_BMI_Plate77WellA11_16S_C5B2R </span></a>
<a class="sourceLine" id="cb8-64" data-line-number="64"><span class="co">#&gt; Sample 1 - 35479 reads in 28363 unique sequences.</span></a>
<a class="sourceLine" id="cb8-65" data-line-number="65"><span class="co">#&gt; Sample 1 - 35479 reads in 26306 unique sequences.</span></a>
<a class="sourceLine" id="cb8-66" data-line-number="66"><span class="co">#&gt; Duplicate sequences in merged output.</span></a>
<a class="sourceLine" id="cb8-67" data-line-number="67"><span class="co">#&gt; Exact sequence variants inferred for sample: runC5B2R_BMI_Plate77WellA11_16S_C5B2R .</span></a>
<a class="sourceLine" id="cb8-68" data-line-number="68"><span class="co">#&gt; Duplicate sequences detected and merged.</span></a>
<a class="sourceLine" id="cb8-69" data-line-number="69"><span class="co">#&gt; Duplicate sequences detected and merged.</span></a>
<a class="sourceLine" id="cb8-70" data-line-number="70"><span class="co">#&gt; Duplicate sequences detected and merged.</span></a>
<a class="sourceLine" id="cb8-71" data-line-number="71"><span class="co">#&gt; Duplicate sequences detected and merged.</span></a>
<a class="sourceLine" id="cb8-72" data-line-number="72"><span class="co">#&gt; Finished processing reads in runC5B2R at 2020-09-09 18:55:44</span></a>
<a class="sourceLine" id="cb8-73" data-line-number="73"><span class="co">#&gt; Sequencing run-specific sequence tables can be found in ~/Documents/UCSC/Projects/NEON_soil_microbe_processing/raw_data/Illumina/NEON/16S/3_seqtabs</span></a></code></pre></div>
<p>Merge the sequence tables from all runs</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="cf">if</span>(<span class="kw">length</span>(all.seqtabs) <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">  seqtab_joined &lt;-<span class="st"> </span>all.seqtabs[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb9-3" data-line-number="3">} <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">  seqtab_joined &lt;-<span class="st"> </span><span class="kw">mergeSequenceTables</span>(<span class="dt">tables =</span> all.seqtabs)</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">}</a></code></pre></div>
<div id="assign-taxonomy-using-the-silva-reference-database" class="section level2">
<h2>Assign taxonomy using the Silva reference database</h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">tax &lt;-<span class="st"> </span><span class="kw">assignTaxonomy</span>(seqtab_joined, SILVA_REF_PATH, <span class="dt">multithread =</span> MULTITHREAD, <span class="dt">verbose =</span> VERBOSE)</a></code></pre></div>
</div>
<div id="save-processed-data-to-file" class="section level2">
<h2>Save processed data to file</h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># Save OTU table and taxonomic table as RDS files</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="co"># to hand off to dada2_to_phyloseq.R</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="kw">saveRDS</span>(seqtab_joined, <span class="st">&quot;./data/NEON_16S_seqtab_nochim.Rds&quot;</span>)</a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="kw">saveRDS</span>(tax, <span class="st">&quot;./data/NEON_16S_tax.Rds&quot;</span>)</a></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
