<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Download NEON Data with Metadata Workaround</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Download NEON Data with Metadata Workaround</h1>



<p>This vignette demonstrates how to use the functions and parameters in this package to download the specific scope NEON soil microbe marker gene sequence data and associated data relevant to your analysis.</p>
<p><strong>This version of of the Download NEON Data vignette does not use the <code>downloadSequenceMetadata()</code> function to download metadata, and instead loads it from memory. This is meant to be a temporary workaround until the NEON API is updated in late 2020 or early 2021 to reflect new naming conventions.</strong></p>
<p><strong>To download the “good” NEON soil microbe sequence metadata, click <a href="https://drive.google.com/file/d/1Rd41yHeHYoNl5HsKWUTO2uFOSLrbcIg9/view?usp=sharing">here</a>. Decompress and move the files into your working directory before running the rest of this vignette.</strong></p>
<p><strong>Note:</strong> If you are using this vignette as a reference, rather than running it as an RMarkdown script, be aware that you will need to manually set the working directory to the project root directory, e.g. <code>setwd(&quot;.../neonSoilMicrobeProcessing&quot;)</code>.</p>
<div id="dependencies" class="section level1">
<h1>Dependencies</h1>
<p>To begin, ensure that you have installed all dependencies.</p>
<p>From CRAN:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">install.packages</span>(<span class="st">&quot;plyr&quot;</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">install.packages</span>(<span class="st">&quot;dplyr&quot;</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">install.packages</span>(<span class="st">&quot;neonUtilities&quot;</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw">install.packages</span>(<span class="st">&quot;R.utils&quot;</span>)</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(plyr)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw">library</span>(neonUtilities)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="kw">library</span>(R.utils)</a></code></pre></div>
<p>In addition, source files associated with this package:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">source</span>(<span class="st">&quot;./R/utils.R&quot;</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">source</span>(<span class="st">&quot;./code/params.R&quot;</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co">#&gt; Could not find the cutadapt tool in file system. This tool is necessary for ITS (fungal) raw sequence processing. Please load the module if necessary on an HPC (i.e. 'module load cutadapt') or download from https://cutadapt.readthedocs.io/en/stable/installation.html</span></a></code></pre></div>
</div>
<div id="set-up-directories" class="section level1">
<h1>Set up directories</h1>
<p>Set up the directory structure associated with the various NEON data products. The structure will depend on what you have written in <code>params.R</code>. For example:</p>
<pre><code>PRESET_OUTDIR = &quot;/data/NEON&quot;
PRESET_OUTDIR_SEQUENCE = &quot;raw_sequence&quot;
PRESET_OUTDIR_SEQMETA = &quot;sequence_metadata&quot;
PRESET_OUTDIR_SOIL = &quot;soil&quot;</code></pre>
<p>This will generate (recursively) the following output directories:</p>
<pre><code>/data/NEON/raw_sequence
/data/NEON/sequence_metadata
/data/NEON/soil</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">makeOutputDirectories</span>()</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co">#&gt; Building output directories from base directory '/Users/lstanish/Github/NEON_soil_microbe_processing/NEON'.</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#&gt; Using '/Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence' as sequence subdirectory.</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">#&gt; Using '/Users/lstanish/Github/NEON_soil_microbe_processing/NEON/sequence_metadata' as sequence metadata subdirectory.</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="co">#&gt; Using '/Users/lstanish/Github/NEON_soil_microbe_processing/NEON/soil' as soil data subdirectory.</span></a></code></pre></div>
</div>
<div id="download-data" class="section level1">
<h1>Download data</h1>
<div id="download-metadata" class="section level2">
<h2>Download metadata</h2>
<p>We begin by downloading the metadata associated with the collection and analysis of NEON soil microbial samples. <strong>In this version of the vignette only, we load metadata from file. In the original version of this vignette, we use the <code>downloadSequenceMetadata()</code> function.</strong></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">meta_ITS &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;./rawFilesGood_ITS.csv&quot;</span>) <span class="co"># For ITS data</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="co"># meta_16S &lt;- read.csv(&quot;./rawFilesGood_16S.csv&quot;) # For 16S data</span></a></code></pre></div>
<p><strong>In the remaining code, we use only the ITS sequence metadata <code>meta_ITS</code>, but the 16S sequence metadata <code>meta_16S</code> can be substituted in order to download and process 16S data instead.</strong></p>
<p>Optionally, subset the metadata to select specific sites, dates, or sequencing runs.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">run_subset_ITS &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CBRDC&quot;</span>)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="co">#run_subset_ITS &lt;- c(&quot;B69PP&quot;, &quot;B69RF&quot;, &quot;B69RN&quot;, &quot;B9994&quot;, &quot;BDR3T&quot;, &quot;BF8M2&quot;, &quot;BF8W6&quot;, &quot;BFDG8&quot;, &quot;BMCBD&quot;, #&quot;BMCC4&quot;, &quot;BNBWL&quot;)</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3">meta &lt;-<span class="st"> </span>meta_ITS[meta_ITS<span class="op">$</span>sequencerRunID <span class="op">%in%</span><span class="st"> </span>run_subset_ITS,]</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="co"># Other options for subsetting by sequencing runs:</span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="co"># run_subset_ITS &lt;- c(&quot;BNM5M&quot;, &quot;BNM6G&quot;, &quot;BNMDC&quot;, &quot;BRGWK&quot;, &quot;BTJKN&quot;, &quot;BTK8C&quot;, &quot;BTNFM&quot;, &quot;BVKHJ&quot;, &quot;BVM74&quot;, &quot;C25G9&quot;, &quot;C25VD&quot;, &quot;C3CN4&quot;)</span></a>
<a class="sourceLine" id="cb8-8" data-line-number="8"><span class="co"># run_subset_ITS &lt;- c(&quot;C3CNF&quot;, &quot;C3D5Y&quot;, &quot;C4M7L&quot;, &quot;C5D7D&quot;, &quot;C5DB5&quot;, &quot;C5DCH&quot;, &quot;C97GV&quot;, &quot;C97PR&quot;, &quot;C9852&quot;, &quot;CBJG3&quot;, &quot;CBRDC&quot;)</span></a>
<a class="sourceLine" id="cb8-9" data-line-number="9"></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="co"># run_subset_16S &lt;- c(&quot;B69PP&quot;, &quot;B69RF&quot;, &quot;B69RN&quot;, &quot;B9994&quot;, &quot;BDNB6&quot;, &quot;BF462&quot;, &quot;BF8M2&quot;, &quot;BFDG8&quot;, &quot;BJ8RK&quot;, &quot;BMC64&quot;, &quot;BMCBJ&quot;)</span></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="co"># run_subset_16S &lt;- c(&quot;BNMJ5&quot;, &quot;BNMWB&quot;, &quot;BRF6G&quot;, &quot;BRPH4&quot;, &quot;BTP4N&quot;, &quot;BTV4F&quot;, &quot;BVKCR&quot;, &quot;BVMB4&quot;, &quot;BVPNR&quot;, &quot;C24VW&quot;, &quot;C25T6&quot;)</span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12"><span class="co"># run_subset_16S &lt;- c(&quot;C38TW&quot;, &quot;C5B2R&quot;, &quot;C5C7P&quot;, &quot;C7WK3&quot;, &quot;C8VMV&quot;, &quot;C977L&quot;, &quot;C983L&quot;, &quot;CBJYB&quot;, &quot;CBTWG&quot;, &quot;CDJ5J&quot;)</span></a></code></pre></div>
</div>
<div id="download-raw-sequence-data" class="section level2">
<h2>Download raw sequence data</h2>
<p>Now that we have the metadata table loaded into memory, we retrieve a table of unique raw data files and their sequencing run IDs.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">download_success &lt;-<span class="st"> </span><span class="kw">downloadRawSequenceData</span>(meta)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="co">#&gt; [1] &quot;There are 5 unique raw sequence files to download.&quot;</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="co">#&gt; Finished downloading /Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/BMI_18S_14_1273_ITS_CBRDC_R2.fastq.gz</span></a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="co">#&gt; Finished downloading /Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/BMI_18S_14_1273_ITS_CBRDC_R1.fastq.gz</span></a>
<a class="sourceLine" id="cb9-5" data-line-number="5"><span class="co">#&gt; Finished downloading /Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/BMI_18S_14_1271_ITS_CBRDC_R1.fastq.gz</span></a>
<a class="sourceLine" id="cb9-6" data-line-number="6"><span class="co">#&gt; Finished downloading /Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/BMI_18S_14_1271_ITS_CBRDC_R2.fastq.gz</span></a>
<a class="sourceLine" id="cb9-7" data-line-number="7"><span class="co">#&gt; Finished downloading /Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/BMI_18S_14_1272_ITS_CBRDC_R2.fastq.gz</span></a></code></pre></div>
</div>
<div id="reorganize-and-rename-the-downloaded-files" class="section level2">
<h2>Reorganize and rename the downloaded files</h2>
<p>Get the names of all downloaded files.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">outdir_sequence &lt;-<span class="st"> </span><span class="kw">file.path</span>(PRESET_OUTDIR, PRESET_OUTDIR_SEQUENCE)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">fn0 &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdir_sequence, meta<span class="op">$</span>rawDataFileName)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">fn &lt;-<span class="st"> </span><span class="kw">unique</span>(fn0[<span class="kw">file.exists</span>(fn0)])</a></code></pre></div>
<p>The following function untars each file (if necessary), appends the sequencer run ID to the beginning of each filename, and moves files to subfolders within your sequence directory according to their target gene. The terminal directory name for either target gene will be <code>0_raw</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">reorganized_files &lt;-<span class="st"> </span><span class="kw">organizeRawSequenceData</span>(fn, meta)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="co">#&gt; Reorganized BMI_18S_14_1273_ITS_CBRDC_R2.fastq.gz</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="co">#&gt; Reorganized BMI_18S_14_1273_ITS_CBRDC_R1.fastq.gz</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="co">#&gt; Reorganized BMI_18S_14_1271_ITS_CBRDC_R1.fastq.gz</span></a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="co">#&gt; Reorganized BMI_18S_14_1271_ITS_CBRDC_R2.fastq.gz</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="co">#&gt; Reorganized BMI_18S_14_1272_ITS_CBRDC_R2.fastq.gz</span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="co">#&gt; 5 out of 5 files in sequencer run CBRDC were successfully reorganized.</span></a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="co">#&gt; Reorganized files can be found in /Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/ITS/0_raw and /Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/16S/0_raw.</span></a>
<a class="sourceLine" id="cb11-9" data-line-number="9"><span class="kw">head</span>(reorganized_files)</a>
<a class="sourceLine" id="cb11-10" data-line-number="10"><span class="co">#&gt; [1] &quot;/Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/ITS/0_raw/runCBRDC_BMI_18S_14_1273_ITS_CBRDC_R2.fastq.gz&quot;</span></a>
<a class="sourceLine" id="cb11-11" data-line-number="11"><span class="co">#&gt; [2] &quot;/Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/ITS/0_raw/runCBRDC_BMI_18S_14_1273_ITS_CBRDC_R1.fastq.gz&quot;</span></a>
<a class="sourceLine" id="cb11-12" data-line-number="12"><span class="co">#&gt; [3] &quot;/Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/ITS/0_raw/runCBRDC_BMI_18S_14_1271_ITS_CBRDC_R1.fastq.gz&quot;</span></a>
<a class="sourceLine" id="cb11-13" data-line-number="13"><span class="co">#&gt; [4] &quot;/Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/ITS/0_raw/runCBRDC_BMI_18S_14_1271_ITS_CBRDC_R2.fastq.gz&quot;</span></a>
<a class="sourceLine" id="cb11-14" data-line-number="14"><span class="co">#&gt; [5] &quot;/Users/lstanish/Github/NEON_soil_microbe_processing/NEON/raw_sequence/ITS/0_raw/runCBRDC_BMI_18S_14_1272_ITS_CBRDC_R2.fastq.gz&quot;</span></a></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
